File: C++/day2/day2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>


int part1(std::vector<std::pair<std::string,int>> instructions)
{
    int height, depth;
    height = 0;
    depth = 0;
    for( std::pair<std::string,int> inst : instructions)
    {
        if(inst.first == "down")
        {
            depth+=inst.second;
        }
        else if(inst.first == "up")
        {
            depth-=inst.second;
        }
        else if(inst.first == "forward")
        {
            height+=inst.second;
        }
    }
    return height * depth;
}

int part2(std::vector<std::pair<std::string,int>> instructions)
{
    int height, depth ,aim;
    height = 0;
    depth = 0;
    aim = 0;
    for( std::pair<std::string,int> instruction : instructions)
    {
        if(instruction.first == "down")
        {
            aim += instruction.second;
        }
        else if(instruction.first == "up")
        {
            aim -= instruction.second;
        }
        else if(instruction.first == "forward")
        {
            height += instruction.second;
            depth += instruction.second * aim;
        }
    }
    return height * depth;

}
int main()
{
    std::vector<std::pair<std::string,int>> instructions;
    std::ifstream input("input.txt");
    if(input.fail()) std::cout << "Failed to open file";
    else {


        while(input.peek() != EOF)
        {
            std::pair<std::string,int> instruction;
            input >> instruction.first >> instruction.second;
            instructions.push_back(instruction);


        }

        std::cout << "Part 1: " << part1(instructions) << std::endl;
        std::cout << "Part 2: " << part2(instructions) << std::endl;
    }

    return 0;
}

File: Python/Day2/day2.py
import os

def part1(instructions):

    depth = 0
    height = 0
    
    for instruction in instructions:

        direction , value = instruction
        
        if direction == "down":
        
            depth += value
        
        elif direction == "up":
        
            depth -= value
        
        elif direction == "forward":
            
            height += value
    
    return depth*height

def part2(instructions):

    depth = 0
    height = 0
    aim = 0
    
    for instruction in instructions:

        direction , value = instruction
        
        if direction == "down":
        
            aim += value
        
        elif direction == "up":
        
            aim -= value
        
        elif direction == "forward":
            
            height += value
            depth += value*aim
    
    return depth*height




def main():
    instructions = []
    
    dirname = os.path.dirname(__file__)
    filename = os.path.join(dirname, 'input.txt')

    with open(filename, encoding='utf8') as f:
        
        for line in f:
            l = line.split(" ")

            instruction = (l[0], int(l[1]))
            instructions.append(instruction)
      
    print("Part one answer : "  + str(part1(instructions)))
    print("Part two answer : "  + str(part2(instructions)))
    
             

if __name__ == "__main__":
    main()
